cache/elasticache_client.go

```diff
@@ -0,0 +1,180 @@
+package cache
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	health "github.com/ONSdigital/dp-healthcheck/healthcheck"
+	"github.com/ONSdigital/dp-sessions-api/session"
+	"github.com/go-redis/redis"
+)
+
+var (
+	HealthyMessage       = "elasticache is OK"
+	ErrEmptySessionID    = errors.New("session id required but was empty")
+	ErrEmptySessionEmail = errors.New("session email required but was empty")
+	ErrEmptySession      = errors.New("session is empty")
+	ErrEmptyAddress      = errors.New("address is empty")
+	ErrEmptyPassword     = errors.New("password is empty")
+	ErrInvalidTTL        = errors.New("ttl should not be zero")
+)
+
+type ElasticacheClient struct {
+	client RedisClienter
+	ttl    time.Duration
+}
+
+// Config - config options for the elasticache client
+type Config struct {
+	Addr     string
+	Password string `json:"-"`
+	Database int
+	TTL      time.Duration
+	TLS      *tls.Config
+}
+
+// New - create new session cache client instance
+func New(c Config) (*ElasticacheClient, error) {
+	if c.Addr == "" {
+		return nil, ErrEmptyAddress
+	}
+
+	if c.Password == "" {
+		return nil, ErrEmptyPassword
+	}
+
+	if c.TTL == 0 {
+		return nil, ErrInvalidTTL
+	}
+
+	return &ElasticacheClient{
+		client: redis.NewClient(&redis.Options{
+			Addr:      c.Addr,
+			Password:  c.Password,
+			DB:        c.Database,
+			TLSConfig: c.TLS,
+		}),
+		ttl: c.TTL,
+	}, nil
+}
+
+// SetSession - add session to elasticache
+func (c *ElasticacheClient) SetSession(s *session.Session) error {
+	if s == nil {
+		return ErrEmptySession
+	}
+
+	sJSON, err := s.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	// Add session using ID as key
+	err = c.client.Set(s.ID, sJSON, c.ttl).Err()
+	if err != nil {
+		return fmt.Errorf("elasticache client.Set returned an unexpected error: %w", err)
+	}
+
+	// Add session using email as key
+	err = c.client.Set(s.Email, sJSON, c.ttl).Err()
+	if err != nil {
+		return fmt.Errorf("elasticache client.Set returned an unexpected error: %w", err)
+	}
+
+	return nil
+}
+
+// GetByID - gets a session from elasticache using its ID
+func (c *ElasticacheClient) GetByID(id string) (*session.Session, error) {
+	if id == "" {
+		return nil, ErrEmptySessionID
+	}
+
+	msg, err := c.client.Get(id).Result()
```

Comment:
How does this client inform you if the requested item is not found? I think we need to specifically check for this case and return a specific error for this scenario as the caller needs to be able to say not found.